File: src/main/java/com/bdev/hogwarts_api/service/events/EventServiceImpl.java
package com.bdev.hogwarts_api.service.events;

import com.bdev.hogwarts_api.dao.EventDao;
import com.bdev.hogwarts_api.data.converter.events.EventDtoConverter;
import com.bdev.hogwarts_api.data.converter.events.EventModelConverter;
import com.bdev.hogwarts_api.data.dto.events.Event;
import com.bdev.hogwarts_api.data.dto.events.EventType;
import com.bdev.hogwarts_api.data.model.events.EventModel;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Optional;

import static java.lang.String.format;
import static java.util.Comparator.comparingLong;
import static java.util.stream.Collectors.toList;

@Service
public class EventServiceImpl implements EventService {
    @Autowired
    private EventDao eventDao;

    @Override
    public List<Event> getAll() {
        return eventDao
                .findAll()
                .stream()
                .map(EventModelConverter::convert)
                .collect(toList());
    }

    @Override
    public Optional<Event> getLatestEvent(EventType eventType) {
        return eventDao
                .getAllByEventTypeAndDateGreaterThan(eventType, new Date().getTime())
                .stream()
                .min(comparingLong(EventModel::getDate))
                .map(EventModelConverter::convert);
    }

    @Override
    public Optional<Event> getById(long id) {
        return Optional
                .ofNullable(eventDao.getOne(id))
                .map(EventModelConverter::convert);
    }

    @Override
    public boolean exists(long id) {
        return eventDao.exists(id);
    }

    @Override
    public long create(Event speakingClub) {
        if (speakingClub.getId() != null) {
            throw new RuntimeException("Event should be null during creation");
        }

        return eventDao.save(EventDtoConverter.convert(speakingClub)).getId();
    }

    @Override
    public void update(Event speakingClub) {
        if (!eventDao.exists(speakingClub.getId())) {
            throw new RuntimeException(format("Event with id '%d' does not exist", speakingClub.getId()));
        }

        eventDao.save(EventDtoConverter.convert(speakingClub));
    }

    @Override
    public void delete(long id) {
        if (!eventDao.exists(id)) {
            throw new RuntimeException(format("Event with id '%d' does not exist", id));
        }

        eventDao.delete(id);
    }
}

File: src/main/java/com/bdev/hogwarts_api/service/events/EventServiceImpl.kt
package com.bdev.hogwarts_api.service.events

import com.bdev.hogwarts_api.dao.EventDao
import com.bdev.hogwarts_api.data.converter.events.EventDtoConverter
import com.bdev.hogwarts_api.data.converter.events.EventModelConverter
import com.bdev.hogwarts_api.data.dto.events.Event
import com.bdev.hogwarts_api.data.dto.events.EventType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.lang.String.format
import java.util.*

@Service
open class EventServiceImpl : EventService {
    @Autowired
    private lateinit var eventDao: EventDao

    override fun getAll(): List<Event> {
        return eventDao
                .findAll()
                .map { EventModelConverter.convert(it) }
    }

    override fun getLatestEvent(eventType: EventType): Event? {
        return eventDao
                .getAllByEventTypeAndDateGreaterThan(eventType, Date().time)
                .minBy { it.date ?: Long.MAX_VALUE }
                ?.let { EventModelConverter.convert(it) }
    }

    override fun getAllFutureEvents(eventType: EventType): List<Event> {
        return eventDao
                .getAllByEventTypeAndDateGreaterThan(eventType, Date().time)
                .map { EventModelConverter.convert(it) }
    }

    override fun getById(id: Long): Event? {
        return eventDao.findOne(id)
                ?.let { EventModelConverter.convert(it) }
    }

    override fun exists(id: Long): Boolean {
        return eventDao.exists(id)
    }

    override fun create(speakingClub: Event): Long {
        if (speakingClub.id != null) {
            throw RuntimeException("Event should be null during creation")
        }

        return eventDao
                .save(EventDtoConverter.convert(speakingClub))
                .id ?: throw RuntimeException()
    }

    override fun update(speakingClub: Event) {
        if (!eventDao.exists(speakingClub.id)) {
            throw RuntimeException(format("Event with id '%d' does not exist", speakingClub.id))
        }

        eventDao.save(EventDtoConverter.convert(speakingClub))
    }

    override fun delete(id: Long) {
        if (!eventDao.exists(id)) {
            throw RuntimeException(format("Event with id '%d' does not exist", id))
        }

        eventDao.delete(id)
    }
}
